db/repair_bucket.go

```diff
@@ -0,0 +1,228 @@
+type DocTransformer func(doc *document) (transformedDoc *document, transformed bool, err error)
+
+// A RepairBucket struct is the main API entrypoint to call for repairing documents in buckets
+type RepairBucket struct {
+	DryRun                  bool // If true, will only output what changes it *would* have made, but not make any changes
+	WriteRepairedDocsToDisk bool // If true, write the before and after doc to temp files
+	Bucket                  base.Bucket
+	RepairJobs              []DocTransformer
+}
+
+func NewRepairBucket(bucket base.Bucket) *RepairBucket {
+	return &RepairBucket{
+		Bucket: bucket,
+	}
+}
+
+func (r *RepairBucket) SetDryRun(dryRun bool) *RepairBucket {
+	r.DryRun = dryRun
+	return r
+}
+
+func (r *RepairBucket) AddRepairJob(repairJob DocTransformer) *RepairBucket {
+	r.RepairJobs = append(r.RepairJobs, repairJob)
+	return r
+}
+
+func (r *RepairBucket) InitFrom(params RepairBucketParams) *RepairBucket {
+
+	r.SetDryRun(params.DryRun)
+	r.WriteRepairedDocsToDisk = true
+	for _, repairJobParams := range params.RepairJobs {
+		switch repairJobParams.RepairJobType {
+		case RepairRevTreeCycles:
+			r.AddRepairJob(RepairJobRevTreeCycles)
+		}
+	}
+
+	return r
+}
+
+func (r RepairBucket) RepairBucket() (repairedDocs []*document, err error) {
+
+	options := Body{"stale": false, "reduce": false}
+	options["startkey"] = []interface{}{true}
+
+	vres, err := r.Bucket.View(DesignDocSyncHousekeeping, ViewImport, options)
+	if err != nil {
+		return repairedDocs, err
+	}
+
+	for _, row := range vres.Rows {
+		rowKey := row.Key.([]interface{})
+		docid := rowKey[1].(string)
+		key := realDocID(docid)
+
+		err = r.Bucket.Update(key, 0, func(currentValue []byte) ([]byte, error) {
+			// Be careful: this block can be invoked multiple times if there are races!
+			if currentValue == nil {
+				return nil, couchbase.UpdateCancel // someone deleted it?!
+			}
+			doc, err := unmarshalDocument(docid, currentValue)
+			if err != nil {
+				return nil, err
+			}
+			updatedDoc, shouldUpdate, err := r.TransformBucketDoc(doc)
+			if err != nil {
+				return nil, err
+			}
+
+			switch shouldUpdate {
+			case true:
+
+				if r.WriteRepairedDocsToDisk {
+					origDocTempFile, writeTempFileErr := writeDocToTempFile(doc)
+					if writeTempFileErr != nil {
+						base.Warn("Error writing orig doc to tmp file.  Doc: %v.  Error: %v", doc, writeTempFileErr)
+					}
+
+					updatedDocTempFile, writeTempFileErr := writeDocToTempFile(updatedDoc)
+					if writeTempFileErr != nil {
+						base.Warn("Error writing update doc to tmp file.  UpdatedDoc: %v.  Error: %v", updatedDoc, writeTempFileErr)
+					}
+
+					base.LogTo("CRUD", "Repair Doc (dry_run=%v).  Original doc written to: %+v, Post-repair doc written to: %+v", r.DryRun, origDocTempFile, updatedDocTempFile)
+
+				} else {
+
+					base.LogTo("CRUD", "Repair Doc (dry_run=%v)  Dumping doc contents directly into logs since WriteRepairedDocsToDisk disabled.", r.DryRun)
+
```

Comment:
I'm not convinced logs are really the best place for full doc bodies + rev trees either - for a repair that touches a lot of documents in the bucket, it's particularly noisy.

Reply:
Fixed.  If insertion of backup doc into bucket fails, it will fallback to dumping to logs.  Should never happen though.